MINIBALL: add Memory virtualization docs

yaoyinnan opened the pull request:
Add description document for Memory virtualization part. Modify the description of Memory virtualization in MiniBall's README.md.

Fixes: https://github.com/openanolis/dragonball-sandbox/issues/200

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
```diff
@@ -0,0 +1,378 @@
+内存虚拟化包含对虚拟机地址空间的管理和对 MMIO 资源的管理，通过 vm-memory 和 dbs-address-space crate 完成上述工作。除此之外，本文在最后介绍了加载内核到 Guest 内存的过程。
+
+## 内存管理
+在配置 KVM 后，开始初始化 guest memory。通过 `vm-memery`、`dbs-address-space` 完成。
```
typo
vm-memory

yaoyinnan replied:
fixed

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
```diff
@@ -0,0 +1,378 @@
+由于port I/O有独立的64K I/O地址空间，但CPU的地址线只有一套，所以必须区分地址属于物理地址空间还是I/O地址空间。
+
+### 虚拟内存
+使用 vm-memery 的 GuestMemoryMmap，可以配置 Guest 的内存空间。`create_guest_memory()` 需要输入从命令行获得的 `MemoryConfig`，包含提供给 Guest 的内存大小 `size_mib`。通过 `create_memory_regions()` 创建内存地址空间，在 x86_64 上，第15 ～ 23 行，如果配置的内存大小超过 MMIO 的起始地址，它们会围绕 MMIO 间隙进行创建（MMIO 设备插槽的专用空间）。
```
typo

yaoyinnan replied:
fixed

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
```diff
@@ -0,0 +1,378 @@
+
+![address3.png](img/address3.png)
+> From [https://zhuanlan.zhihu.com/p/69828213](https://zhuanlan.zhihu.com/p/69828213)
+
```
Could you add a pic here to demonstrate traditional guest memory layout?

yaoyinnan replied:
added

studychao reviewed (commented):
Thanks yinnan, a few comments...

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
```diff
@@ -0,0 +1,381 @@
+
+第 9 ～ 11 行，先判断输入的MMIO 起始地址是否小于终止地址。第 13 ～ 22 行，构造启动参数，其中第 20 ～ 22 行，添加内核命令行地址到启动参数中。第 24 ～ 26 行，通过 `add_e820_entry()` 方法将启动参数信息加载到数据区 EBDA 中。第 28 ～ 69 行，根据 Guest 内存的大小与 MMIO 地址范围的关系将启动参数写入对应的位置。
+第 72 ～ 79 行，将启动参数写入 zeropage 中。
+> zero page是一个特殊的物理页，里面值全部为0，zero page是针对匿名页场景专门进行优化，主要是节省内存和对性能进行了一定优化。当malloc或者mapp一段虚拟内存后，第一次对该内存访问为读操作，将会发生匿名page fault。do_anonymous_page处理，由于第一次为读操作还未发生写操作，因此发生一个zero page，为其申请一个特殊物理页zero page。
```
The zero page in your code should be referred to this concept (https://www.kernel.org/doc/html/v5.8/x86/zero-page.html) in linux kernel.

yaoyinnan replied:
I re-described it with reference to the Linux kernel docs.

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
````diff
@@ -0,0 +1,381 @@
+```
+
+#### 配置启动参数
+第 6 ～ 16 行，配置启动参数，参数的详细内容在之后代码中介绍。
````
Where is the detailed introduction?

yaoyinnan replied:
added

yaoyinnan replied:
Sorry, I am still implementing virtio net in Miniball, and I will supplement the description in this part later.

yaoyinnan replied:
Sorry, I am still implementing virtio net in Miniball, and I will supplement the description in this part after I implementing virtio net.

studychao commented on crates/dbs-miniball/docs/MemoryVirtualization.md:
```diff
@@ -0,0 +1,408 @@
+## 将 guest kernel 加载到 guest memory 中
+将 guest kernel 加载到 guest memory 中。通过 `linux-loader` 和 `dbs-boot` crate 完成。
+
+1. linux-loader：支持在 x86_64 上加载原始 ELF (vmlinux) 和压缩的大 zImage (bzImage) 格式内核映像。
```
大zImage的说法有点奇怪，因为bzImage意思是big zImage，翻译了其中一部分留了另外一部分不太好

yaoyinnan replied:
sorry, I allredy fix it.

studychao commented:
please upgrade the memory layout graph

yaoyinnan committed:
MINIBALL: add Memory virtualization docs

Add description document for Memory virtualization part. Modify the description of Memory virtualization in MiniBall's README.md.

Fixes: https://github.com/openanolis/dragonball-sandbox/issues/200

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>
